File: 92-ReverseLinkedListIi/92-ReverseLinkedListIi.py
# Last updated: 12/6/2025, 5:53:29 am
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        node = head
        n = right-left
        mid = n // 2
        for i in range(1, left): node = node.next
        def rec(node, index):
            nonlocal left
            if index > n: return
            rec(node.next, index+1)
            if index > mid:
                tmp = node.val
                node.val = left.val
                left.val = tmp
                left = left.next
        left = node
        rec(node, 0)
        return head

File: 92-ReverseLinkedListIi/92-ReverseLinkedListIi.cpp
// Last updated: 12/6/2025, 5:53:35 am
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void rec(ListNode *node, int index, int &left, int &right, unordered_map<int, int> &hash){
        if (index > right) return;
        hash[left+(right-index)] = node->val;
        rec(node->next, index+1, left, right, hash);
        node->val = hash[index];
    }
    
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        unordered_map<int, int> hash;
        ListNode* node = head;
        for (int i=1; i<left; i++) node = node->next;
        rec(node, left, left, right, hash);
        return head;
    }
};